### PR TITLE
Menubar settings

### PR DESCRIPTION
**Pull request by pjan**
As per title.

### CODE REVIEW

**Comment by pjan** [comment]
What can be done in a follow-up PR:
- enable/disable transparency in the menu
- whether to reveal IP address, hostname, OS version, etc. when clicking the clock in the menubar in the login window

**Comment by cbarrett** [comment]
How does this affect (if at all) things using the public [NSStatusItem](https://developer.apple.com/documentation/appkit/nsstatusitem) api?

File: modules/system/menubar.nix [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    WWAN            = "/System/Library/CoreServices/Menu Extras/WWAN.menu";
+    iChat           = "/System/Library/CoreServices/Menu Extras/iChat.menu";
+  };
+
+in {
+  options = {
+    system.menubar.items = mkOption {
+      type = types.nullOr (types.listOf (types.enum (attrNames items)));
+      default = null;
+      description = "Items to display in the menubar";
```
Comment (LnL7) [inline modules/system/menubar.nix]: Could you add an example and preferably a description of the default value.

File: modules/system/menubar.nix [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    };
+  };
+
+  config = {
+
+    system.activationScripts.menubar.text = optionalString (!isNull cfg.items) ''
+      defaults write com.apple.systemuiserver menuExtras -array ${builtins.toString (map (k: "'" + (builtins.getAttr k items) + "'") cfg.items)}
```
Comment (LnL7) [inline modules/system/menubar.nix]: I would prefer to use the defaults module as the implementation for this.

**Comment by pjan** [comment]
> How does this affect (if at all) things using the public NSStatusItem api?
It doesn't

> This looks like an option that should revert it's activation when unset.
What do you mean? What's the action point?

File: modules/system/menubar.nix (hunk 2 above)
Comment (pjan) [reply modules/system/menubar.nix]: There's other menubar settings that can be set, not all fall under the `defaults` module, so I feel that would be more messy for end-users.

I have similar feelings with some of the existing settings in the `defaults` module; e.g. settings for "Printing" -> we now have the "show expanded dialog" option, but we could also set "automatic closing of the window when printing completed", which would then be in a different config namespace. I feel the api would be much cleaner if you would have all `printing` settings together. WDYT?

Comment (LnL7) [reply modules/system/menubar.nix]: To clarify I'm only referring to the activation scripts, not the `system.menubar` option. Modules can set options for other modules as their implementation. Something like this separates the details of defaults write from the rest of the logic.

```nix
# ...
{
  options = {
    system.menubar.items = mkOption {
      type = types.listOf (types.enum (attrValues itemPaths));
      default = [];
      example = [ "Airport" ];
    };
  };
  config = {
    system.defaults.systemuiserver.menuExtras = mkIf (cfg.items != []) (map (k: getAttr k itemPaths) cfg.items);
  };
}
```

**Comment by LnL7** [comment]
Unlike services/etc, defaults don't change back to their original state when removed. To revert back you have to either rebuild with the default value mentioned in the description or use the ui.

I'm just wondering if that's ok for this.

File: modules/system/menubar.nix (hunk 2 above)
Comment (LnL7) [reply modules/system/menubar.nix]: Does that make sense?

**Comment by LnL7** [comment]
Ping

File: modules/system/defaults/systemUIServer.nix [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    WWAN            = "/System/Library/CoreServices/Menu Extras/WWAN.menu";
+    iChat           = "/System/Library/CoreServices/Menu Extras/iChat.menu";
+  };
+
+in {
+
+  options = {
+    system.defaults.systemUIServer.menuExtras = mkOption {
+      type = types.nullOr (types.listOf (types.enum (attrNames items)));
+      example = [ "Airport" "Battery" "Bleutooth" ];
```
Comment (LnL7) [inline modules/system/defaults/systemUIServer.nix]: typo

File: modules/system/defaults-write.nix [hunk 4]
```diff
@@ -58,5 +59,10 @@ in
         ${concatStringsSep "\n" trackpadBluetooth}
```
Comment (LnL7) [inline modules/system/defaults-write.nix]: `systemUIServer` entry is missing

**Comment by LnL7** [comment]
What version of macOS are you using? Setting the defaults write seems to get overridden after restarting the SystemUIServer process.

**Pull request closed without merging**